Latency measurement (for opmon) throughout trigger flow

MRiganSUSX opened the pull request:
This PR addresses the addition of latency measurements at multiple sampling points throughout the trigger.

The general idea is to compare the difference between the current system time (using chrono clock) and the data time (timestamp of the TX objects). There is also an initial offset applied (only calculated once per module), to make this also functional with offline/replay.
The time differences are then stored to opmon variables (opmon plots to be added soon). The idea is to have a monitoring of latency throughout trigger. We can see whether the requests we make are sensible (ie we are not requesting data many seconds in history), and we can also see what 'steps' in the flow contribute the most to the latency (this is slightly obscured due to the offset, something to improve in time).

There are also associated PRs: 
https://github.com/DUNE-DAQ/triggeralgs/pull/69
https://github.com/DUNE-DAQ/daqconf/pull/461

Current sampling points:
-> TPSets entering TPChannelFilter (first plugin in the trigger flow)
-> TX objects entering AND leaving any maker: this is a part of the TriggerGenericMaker which is the class used by both Activity makers and Candidate makers; the actual implementation (updating) depends on particular maker though
-> TCs entering MLT
-> at the point when TD is made for a particular TC
-> at the point when TD is ready and being sent to DFO for a particular TC

(It is possible there will be improvements to the logic, and/or additions/removals of sampling points).

Sampling points (red arrows) shown on the flow chart:
![flow](https://github.com/DUNE-DAQ/trigger/assets/32065363/a58d4b70-cc3c-4b7c-b11e-66b45bb979a2)

Names for new opmon variables:

- TPChannelFilter: 
  - data_vs_system_time_ms
- TriggerGenericMaker: 
  - data_vs_system_in_ms
  - data_vs_system_out_ms
- ModuleLevelTrigger:
  - tc_data_vs_system_ms
  - td_made_vs_ro_window_ms
  - td_send_vs_ro_window_ms

Few examples from the info_trigger.json file:
```
{"__parent":{"mrigan-test.trigger":{"__children":{"chain_input_0":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372276}}},"channelfilter_tplink1":{"__properties":{"dunedaq.trigger.tpchannelfilterinfo.Info":{"__data":{"data_vs_system_time_ms":63,"received_count":4329,"sent_count":4329},"__time":1715372276}}},"channelfilter_tplink1.tpset_sink_to_tpsettee_tplink1.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372276}}},"data_requests_for_Trigger_0x00000001":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":700,"messages":5132},"__time":1715372276}}},"data_requests_for_Trigger_0x00000002":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":700,"messages":827},"__time":1715372276}}},"data_requests_for_Trigger_0x00000003":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":700,"messages":837},"__time":1715372276}}},"df_busy_signal":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":0,"messages":7663},"__time":1715372276}}},"fragments_to_dataflow0":{"__properties":{"dunedaq.ipm.senderinfo.Info":{"__data":{"bytes":53963,"messages":30},"__time":1715372276}}},"mlt":{"__properties":{"dunedaq.trigger.moduleleveltriggerinfo.Info":{"__data":{"lc_kDead":0,"lc_kLive":7944,"lc_kPaused":157,"new_td_inhibited_count":0,"new_td_sent_count":10,"new_td_total_count":10,"tc_data_vs_system_ms":1969,"tc_ignored_count":0,"tc_received_count":22,"td_cleared_count":0,"td_cleared_tc_count":0,"td_dropped_count":0,"td_dropped_tc_count":0,"td_inhibited_count":0,"td_inhibited_tc_count":0,"td_made_vs_ro_window_ms":1969,"td_not_triggered_count":0,"td_not_triggered_tc_count":0,"td_paused_count":0,"td_paused_tc_count":0,"td_queue_timeout_expired_err_count":0,"td_queue_timeout_expired_err_tc_count":0,"td_send_vs_ro_window_ms":2070,"td_sent_count":10,"td_sent_tc_count":22,"td_total_count":10},"__time":1715372276}}},"tam_1_0":{"__properties":{"dunedaq.trigger.triggergenericmakerinfo.Info":{"__data":{"data_vs_system_in_ms":59,"data_vs_system_out_ms":430,"received_count":4331,"sent_count":12},"__time":1715372276}}},"tam_1_0.output_to_tasettee_region_1_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372276}}},"tas0_to_tazipper_0":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372276}}},"tasettee_region_1_0.output2_to_ta_buf_region_1.taset_source":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372276}}},"tazipper_0":{"__properties":{"dunedaq.trigger.triggerzipperinfo.Info":{"__data":{"n_received":12,"n_sent":11,"n_tardy":0},"__time":1715372276}}},"tazipper_0.output_to_tcm_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372276}}},"tcm_0":{"__properties":{"dunedaq.trigger.triggergenericmakerinfo.Info":{"__data":{"data_vs_system_in_ms":1970,"data_vs_system_out_ms":1970,"received_count":11,"sent_count":22},"__time":1715372276}}},"tcs_to_buf":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372276}}},"tcs_to_mlt":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372276}}},"td_to_dfo":{"__properties":{"dunedaq.ipm.senderinfo.Info":{"__data":{"bytes":850,"messages":10},"__time":1715372276}}},"tpsets_tplink0":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":1404056,"messages":4329},"__time":1715372276}}},"tpsettee_tplink1.output1_to_tam_1_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372276}}},"tpsettee_tplink1.output2_to_buf_tplink1.tpset_source":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372276}}}},"__properties":{"dunedaq.appfwk.appinfo.Info":{"__data":{"busy":false,"error":false,"host":"np04-srv-016","state":"RUNNING"},"__time":1715372276},"dunedaq.rcif.runinfo.Info":{"__data":{"running":true,"runno":1,"runtime":8},"__time":1715372276}}}},"__tags":{"partition_id":"mrigan-test"}}
{"__parent":{"mrigan-test.trigger":{"__children":{"chain_input_0":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372286}}},"channelfilter_tplink1":{"__properties":{"dunedaq.trigger.tpchannelfilterinfo.Info":{"__data":{"data_vs_system_time_ms":140,"received_count":9752,"sent_count":9751},"__time":1715372286}}},"channelfilter_tplink1.tpset_sink_to_tpsettee_tplink1.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372286}}},"data_requests_for_Trigger_0x00000001":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":910,"messages":6413},"__time":1715372286}}},"data_requests_for_Trigger_0x00000002":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":910,"messages":1021},"__time":1715372286}}},"data_requests_for_Trigger_0x00000003":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":910,"messages":1023},"__time":1715372286}}},"df_busy_signal":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":0,"messages":9473},"__time":1715372286}}},"fragments_to_dataflow0":{"__properties":{"dunedaq.ipm.senderinfo.Info":{"__data":{"bytes":26248,"messages":39},"__time":1715372286}}},"mlt":{"__properties":{"dunedaq.trigger.moduleleveltriggerinfo.Info":{"__data":{"lc_kDead":0,"lc_kLive":17951,"lc_kPaused":157,"new_td_inhibited_count":0,"new_td_sent_count":13,"new_td_total_count":13,"tc_data_vs_system_ms":1825,"tc_ignored_count":0,"tc_received_count":38,"td_cleared_count":0,"td_cleared_tc_count":0,"td_dropped_count":0,"td_dropped_tc_count":0,"td_inhibited_count":0,"td_inhibited_tc_count":0,"td_made_vs_ro_window_ms":1825,"td_not_triggered_count":0,"td_not_triggered_tc_count":0,"td_paused_count":0,"td_paused_tc_count":0,"td_queue_timeout_expired_err_count":0,"td_queue_timeout_expired_err_tc_count":0,"td_send_vs_ro_window_ms":1925,"td_sent_count":23,"td_sent_tc_count":38,"td_total_count":23},"__time":1715372286}}},"tam_1_0":{"__properties":{"dunedaq.trigger.triggergenericmakerinfo.Info":{"__data":{"data_vs_system_in_ms":135,"data_vs_system_out_ms":972,"received_count":9751,"sent_count":25},"__time":1715372286}}},"tam_1_0.output_to_tasettee_region_1_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372286}}},"tas0_to_tazipper_0":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372286}}},"tasettee_region_1_0.output2_to_ta_buf_region_1.taset_source":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372286}}},"tazipper_0":{"__properties":{"dunedaq.trigger.triggerzipperinfo.Info":{"__data":{"n_received":25,"n_sent":24,"n_tardy":0},"__time":1715372286}}},"tazipper_0.output_to_tcm_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372286}}},"tcm_0":{"__properties":{"dunedaq.trigger.triggergenericmakerinfo.Info":{"__data":{"data_vs_system_in_ms":1826,"data_vs_system_out_ms":1826,"received_count":24,"sent_count":38},"__time":1715372286}}},"tcs_to_buf":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372286}}},"tcs_to_mlt":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372286}}},"td_to_dfo":{"__properties":{"dunedaq.ipm.senderinfo.Info":{"__data":{"bytes":1105,"messages":13},"__time":1715372286}}},"tpsets_tplink0":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":1897521,"messages":5423},"__time":1715372286}}},"tpsettee_tplink1.output1_to_tam_1_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372286}}},"tpsettee_tplink1.output2_to_buf_tplink1.tpset_source":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1715372286}}}},"__properties":{"dunedaq.appfwk.appinfo.Info":{"__data":{"busy":false,"error":false,"host":"np04-srv-016","state":"RUNNING"},"__time":1715372286},"dunedaq.rcif.runinfo.Info":{"__data":{"running":true,"runno":1,"runtime":18},"__time":1715372286}}}},"__tags":{"partition_id":"mrigan-test"}}
```

Note: this is currently only implemented in non-debug algos that we are expected to be using: ChannelAdjacency, ChannelDistance, HorizontalMuon, Prescale.

MRiganSUSX committed:
latency measurements stored to opmon

MRiganSUSX committed:
more changes re latency measurements

ArturSztuc commented on plugins/ModuleLevelTrigger.cpp:
```diff
@@ -194,6 +201,13 @@ ModuleLevelTrigger::do_start(const nlohmann::json& startobj)
+  m_td_send_vs_ro = 0;
+  m_first_tc = true;
+  m_initial_offset = 0;
+
```
I think this probably should be the top of `do_start`, so it's before e.g. the  trigger-decision sending threads? Or even in the constructor.

MRiganSUSX replied:
moved

ArturSztuc commented on plugins/ModuleLevelTrigger.cpp:
```diff
@@ -330,6 +344,16 @@ ModuleLevelTrigger::send_trigger_decisions()
     std::optional<triggeralgs::TriggerCandidate> tc = m_candidate_input->try_receive(std::chrono::milliseconds(10));
     if (tc.has_value()) {
+
+      if (m_first_tc) {
+        m_initial_offset = (duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count()) - (tc->time_start*(16*1e-6));
+        m_first_tc = false;
+      }
```
Just in case, was this tested with the replay and with some real detector run (I guess warmbox)?

MRiganSUSX replied:
tested with the default file from instructions page and with replay, not with real detector yet (can only be tested once in patch, next week)

ArturSztuc commented on plugins/ModuleLevelTrigger.cpp:
```diff
@@ -330,6 +344,16 @@ ModuleLevelTrigger::send_trigger_decisions()
+
+      // Update OpMon Variable(s)
+      uint64_t system_time = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+      m_tc_data_vs_system.store(fabs(system_time - tc->time_start*(16*1e-6) - m_initial_offset));
```
I guess the `16*1e-6` is not defined anywhere in the global configuration / appfwk / outside trigger where we could reasonably take it from? 
I imagine not, and if not, then maybe we should have a private `const` data member `m_` somewhere in the header with this number in (and a comment explaining what it is), and use that private data member instead of having magic numbers everywhere :/

MRiganSUSX replied:
I'm not aware of this being anywhere in global config, but there are other occurrences of these translations elsewhere in the code. For now moved to header(s).

ArturSztuc commented on plugins/ModuleLevelTrigger.cpp:
```diff
@@ -457,11 +482,15 @@ ModuleLevelTrigger::call_tc_decision(const ModuleLevelTrigger::PendingTD& pendin
+
+        // block to update latency TD send vs readout time window start
+        uint64_t system_time = std::chrono::duration_cast<std::chrono::milliseconds>(system_clock::now().time_since_epoch()).count();
+        m_td_send_vs_ro.store( system_time - decision.components.front().window_begin*(16*1e-6) - m_initial_offset );
```
I'm surprised this works here with the `std::move` above it (maybe this should be above the `td_sender->send` line?). 
Either way, would be good to also the `window_end` also, so we know whether the entire window is e.g. well past our buffer etc. E.g. if most of our TDs will be 1ms long, then that's fine, if a lot of them will be e.g. 1s, then it's something to track jic

MRiganSUSX replied:
It did work, but moved before the line for safety (I originally only wanted to report on the values after the try move to not slow it down)

ArturSztuc commented on plugins/ModuleLevelTrigger.hpp:
```diff
@@ -224,6 +224,13 @@ class ModuleLevelTrigger : public dunedaq::appfwk::DAQModule
   std::atomic<metric_counter_type> m_lc_kLive{ 0 };
   std::atomic<metric_counter_type> m_lc_kPaused{ 0 };
   std::atomic<metric_counter_type> m_lc_kDead{ 0 };
+  std::atomic<uint64_t>            m_tc_data_vs_system;
+  std::atomic<uint64_t>            m_td_made_vs_ro;
+  std::atomic<uint64_t>            m_td_send_vs_ro;
+
+  // Latency
+  std::atomic<bool>                m_first_tc;
+  std::atomic<uint64_t>            m_initial_offset;
```
I don't quite understand why these need to be atomic? MLT has only one thread, `m_send_trigger_decisions_thread`, so what else do we need to synchronise at the time measurement point? Atomics are normally used when memory is shared between multiple running threads, but only one thread is using these?

ArturSztuc replied:
I was wrong here, resolved. I now see there are at least two threads using these.

ArturSztuc commented on plugins/TPChannelFilter.cpp:
```diff
@@ -136,6 +140,17 @@ TPChannelFilter::do_work()
     if (tpset->type == TPSet::kPayload) {
+
+      if (m_first_tp) {
+        m_initial_offset = (duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count()) - (tpset->objects.begin()->time_start*(16*1e-6));
+        m_first_tp = false;
```
Ditto with the `16*1e-6`. I wonder if it should be hard-coded somewhere in the configuration instead...

MRiganSUSX replied:
as above, now moved to header(s)

ArturSztuc commented on plugins/TPChannelFilter.hpp:
```diff
@@ -55,6 +55,9 @@ class TPChannelFilter : public dunedaq::appfwk::DAQModule
   using metric_counter_type = decltype(tpchannelfilterinfo::Info::received_count);
   std::atomic<metric_counter_type> m_received_count;
   std::atomic<metric_counter_type> m_sent_count;
+  std::atomic<uint64_t>            m_data_vs_system_time;
+  std::atomic<uint64_t>            m_initial_offset;
+  std::atomic<bool>                m_first_tp;
```
Ditto w.r.t. the threads, are these variables shared between different threads?

ArturSztuc replied:
I was wrong here, resolved.

ArturSztuc reviewed (changes requested):
Overall LGTM, there are a few things that I found odd though, mostly to do with use of `std::atomic`, magic numbers (somewhat unavoidable), and that it would be nice to have the `decision.window_end` measurement too, so we track both sides of the window.

Could you also write how you tested it? Replay and a live-run with some APAs?

On top of this, I would like to tag @glehmannmiotto here, just for tracking: this is an important feature for PD2, and I think we would like this in np02/v5 too. Some of it will be easy-ish to port, some maybe not so much, and we might need to implement such latency measurements in ReadoutModel too (if not already there and I just didn't look enough) before NP02. I hope the trigger team will be able to start working on this in a month or so, but we'll see...

ArturSztuc commented:
Huh, ignore all the `std::atomic` comments. I see this now, there are at least 2 threads using these variables

MRiganSUSX committed:
making latency offset configurable

MRiganSUSX commented:
Several changes:

- addressed your comments:
  - moved few pieces of code around (to 'safer' places)
  - moved 'magic number' (translation from ticks to ms) to header(s), where appropriate
  - added the additional opmon for MLT measuring the window_end difference to system time
  
- new features:
  - aforementioned new opmon in MLT wrt window_end of TD
  - the latency_offset is now configurable, off by default: therefore for live running we do not need/want this, but it is still possible to use it for offline/replay. Same flag is used (one config parameter in trigger json) and passed to channel filter, activity/candidate maker, mlt; for consistency throughout. The implementation is in the base classes (use_latency function), so there are no issues if the updates are not explicitly implemented in specific algos. 
  - updated across the algos (prescale, HM, CHA, CHD)

ArturSztuc committed:
Removing TP buffers from the pluggins

ArturSztuc committed:
Removing any mention of TPBuffer from cmake & readme

MRiganSUSX committed:
second round of latency monitoring updates

MRiganSUSX committed:
second round of latency monitoring updates

MRiganSUSX committed:
another round of latency monitoring updates

MRiganSUSX committed:
clang format

MRiganSUSX commented:
Another version, updates:

- the use of the latency reporting is configurable:
```
        "enable_latency_monitoring": false,
        "use_latency_offset": false
```
`enable_latency_monitoring` -> overall latency reporting on/off
`use_latency_offset` -> offset to apply if running with offline data/replay
both are off (false) by default
- the reporting was moved to `TriggerGenericMaker`, so it is available for every single algo (this is the base class the makers use) and does not require explicit updating in each algo. This also means that the rates are 'more manageable' because they are reported per set (instead of each data object). This required a 'hack' because the generic maker does not have a configuration logic. Therefore, the latency flags are stashed inside the maker configuration block. 
- it's still possible to assign the info reporting given the maker name (ie tam_1, tcm_0, etc)
- other sampling points are still in the `TPChannelFiler` and in the `MLT`
- also applied clang formatting so it may seem like many changes (especially for algos), but outside of the latency reporting there aren't many

the linked PRs (still):
- triggeralgs: https://github.com/DUNE-DAQ/triggeralgs/pull/69/files
- daqconf: https://github.com/DUNE-DAQ/daqconf/pull/461/files

I have only tested with the default data file (with prescale and ADC SW) and with replay.
Passes all integration tests. 
More testing to follow. Should look out for the impact on performance.
STILL NEEDS GRAFANA PLOTS. 

Example info output:
```
{"__parent":{"mrigan-test.trigger":{"__children":{"chain_input_0":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497157}}},"channelfilter_tplink1":{"__properties":{"dunedaq.trigger.tpchannelfilterinfo.Info":{"__data":{"data_vs_system_time_ms":20,"received_count":84337,"sent_count":84337},"__time":1717497157}}},"channelfilter_tplink1.tpset_sink_to_tpsettee_tplink1.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497157}}},"data_requests_for_Trigger_0x00000001":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":88704,"messages":18480},"__time":1717497157}}},"data_requests_for_Trigger_0x00000002":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":88704,"messages":2407},"__time":1717497157}}},"data_requests_for_Trigger_0x00000003":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":88632,"messages":2403},"__time":1717497157}}},"df_busy_signal":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":0,"messages":9455},"__time":1717497157}}},"fake_hsievents":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":150,"messages":9464},"__time":1717497157}}},"fragments_to_dataflow0":{"__properties":{"dunedaq.ipm.senderinfo.Info":{"__data":{"bytes":756909,"messages":3695},"__time":1717497157}}},"mlt":{"__properties":{"dunedaq.trigger.moduleleveltriggerinfo.Info":{"__data":{"lc_kDead":648,"lc_kLive":47301,"lc_kPaused":199,"new_td_inhibited_count":0,"new_td_sent_count":1231,"new_td_total_count":1231,"tc_data_vs_system_ms":0,"tc_ignored_count":0,"tc_received_count":5915,"td_cleared_count":0,"td_cleared_tc_count":0,"td_dropped_count":0,"td_dropped_tc_count":0,"td_inhibited_count":79,"td_inhibited_tc_count":79,"td_made_vs_ro_window_ms":0,"td_not_triggered_count":0,"td_not_triggered_tc_count":0,"td_paused_count":3,"td_paused_tc_count":3,"td_queue_timeout_expired_err_count":0,"td_queue_timeout_expired_err_tc_count":0,"td_send_vs_ro_end_ms":0,"td_send_vs_ro_start_ms":0,"td_sent_count":5818,"td_sent_tc_count":5819,"td_total_count":5900},"__time":1717497157}}},"tam_1_0":{"__properties":{"dunedaq.trigger.triggergenericmakerinfo.Info":{"__data":{"data_vs_system_in_ms":21,"data_vs_system_out_ms":38,"received_count":84337,"sent_count":5868},"__time":1717497157}}},"tam_1_0.output_to_tasettee_region_1_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497157}}},"tas0_to_tazipper_0":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497157}}},"tasettee_region_1_0.output2_to_ta_buf_region_1.taset_source":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497157}}},"tazipper_0":{"__properties":{"dunedaq.trigger.triggerzipperinfo.Info":{"__data":{"n_received":5868,"n_sent":5867,"n_tardy":0},"__time":1717497157}}},"tazipper_0.output_to_tcm_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497157}}},"tcm_0":{"__properties":{"dunedaq.trigger.triggergenericmakerinfo.Info":{"__data":{"data_vs_system_in_ms":23,"data_vs_system_out_ms":23,"received_count":5867,"sent_count":5866},"__time":1717497157}}},"tcs_to_buf":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497157}}},"tcs_to_mlt":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497157}}},"td_to_dfo":{"__properties":{"dunedaq.ipm.senderinfo.Info":{"__data":{"bytes":220349,"messages":1231},"__time":1717497157}}},"tpsets_tplink0":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":4674965,"messages":17497},"__time":1717497157}}},"tpsettee_tplink1.output1_to_tam_1_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497157}}},"tpsettee_tplink1.output2_to_buf_tplink1.tpset_source":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":6},"__time":1717497157}}},"ttcm_fake":{"__properties":{"dunedaq.trigger.timingtriggercandidatemakerinfo.Info":{"__data":{"tc_sent_count":49,"tc_sig_type_err_count":0,"tc_total_count":49,"tsd_received_count":49},"__time":1717497157}}},"ttcm_fake_input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497157}}}},"__properties":{"dunedaq.appfwk.appinfo.Info":{"__data":{"busy":false,"error":false,"host":"np04-srv-016","state":"RUNNING"},"__time":1717497157},"dunedaq.rcif.runinfo.Info":{"__data":{"running":true,"runno":42,"runtime":48},"__time":1717497157}}}},"__tags":{"partition_id":"mrigan-test"}}
{"__parent":{"mrigan-test.trigger":{"__children":{"chain_input_0":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497167}}},"channelfilter_tplink1":{"__properties":{"dunedaq.trigger.tpchannelfilterinfo.Info":{"__data":{"data_vs_system_time_ms":20,"received_count":101878,"sent_count":101878},"__time":1717497167}}},"channelfilter_tplink1.tpset_sink_to_tpsettee_tplink1.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497167}}},"data_requests_for_Trigger_0x00000001":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":88632,"messages":18523},"__time":1717497167}}},"data_requests_for_Trigger_0x00000002":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":88632,"messages":2391},"__time":1717497167}}},"data_requests_for_Trigger_0x00000003":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":88704,"messages":2406},"__time":1717497167}}},"df_busy_signal":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":0,"messages":9458},"__time":1717497167}}},"fake_hsievents":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":150,"messages":9468},"__time":1717497167}}},"fragments_to_dataflow0":{"__properties":{"dunedaq.ipm.senderinfo.Info":{"__data":{"bytes":756714,"messages":3694},"__time":1717497167}}},"mlt":{"__properties":{"dunedaq.trigger.moduleleveltriggerinfo.Info":{"__data":{"lc_kDead":648,"lc_kLive":57307,"lc_kPaused":199,"new_td_inhibited_count":0,"new_td_sent_count":1232,"new_td_total_count":1232,"tc_data_vs_system_ms":0,"tc_ignored_count":0,"tc_received_count":7147,"td_cleared_count":0,"td_cleared_tc_count":0,"td_dropped_count":0,"td_dropped_tc_count":0,"td_inhibited_count":79,"td_inhibited_tc_count":79,"td_made_vs_ro_window_ms":0,"td_not_triggered_count":0,"td_not_triggered_tc_count":0,"td_paused_count":3,"td_paused_tc_count":3,"td_queue_timeout_expired_err_count":0,"td_queue_timeout_expired_err_tc_count":0,"td_send_vs_ro_end_ms":0,"td_send_vs_ro_start_ms":0,"td_sent_count":7050,"td_sent_tc_count":7051,"td_total_count":7132},"__time":1717497167}}},"tam_1_0":{"__properties":{"dunedaq.trigger.triggergenericmakerinfo.Info":{"__data":{"data_vs_system_in_ms":21,"data_vs_system_out_ms":38,"received_count":101878,"sent_count":7090},"__time":1717497167}}},"tam_1_0.output_to_tasettee_region_1_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497167}}},"tas0_to_tazipper_0":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497167}}},"tasettee_region_1_0.output2_to_ta_buf_region_1.taset_source":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":1},"__time":1717497167}}},"tazipper_0":{"__properties":{"dunedaq.trigger.triggerzipperinfo.Info":{"__data":{"n_received":7090,"n_sent":7089,"n_tardy":0},"__time":1717497167}}},"tazipper_0.output_to_tcm_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497167}}},"tcm_0":{"__properties":{"dunedaq.trigger.triggergenericmakerinfo.Info":{"__data":{"data_vs_system_in_ms":22,"data_vs_system_out_ms":22,"received_count":7089,"sent_count":7088},"__time":1717497167}}},"tcs_to_buf":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497167}}},"tcs_to_mlt":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497167}}},"td_to_dfo":{"__properties":{"dunedaq.ipm.senderinfo.Info":{"__data":{"bytes":220528,"messages":1232},"__time":1717497167}}},"tpsets_tplink0":{"__properties":{"dunedaq.ipm.receiverinfo.Info":{"__data":{"bytes":4676204,"messages":17541},"__time":1717497167}}},"tpsettee_tplink1.output1_to_tam_1_0.input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497167}}},"tpsettee_tplink1.output2_to_buf_tplink1.tpset_source":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":6},"__time":1717497167}}},"ttcm_fake":{"__properties":{"dunedaq.trigger.timingtriggercandidatemakerinfo.Info":{"__data":{"tc_sent_count":59,"tc_sig_type_err_count":0,"tc_total_count":59,"tsd_received_count":59},"__time":1717497167}}},"ttcm_fake_input":{"__properties":{"dunedaq.iomanager.queueinfo.Info":{"__data":{"capacity":1000,"number_of_elements":0},"__time":1717497167}}}},"__properties":{"dunedaq.appfwk.appinfo.Info":{"__data":{"busy":false,"error":false,"host":"np04-srv-016","state":"RUNNING"},"__time":1717497167},"dunedaq.rcif.runinfo.Info":{"__data":{"running":true,"runno":42,"runtime":58},"__time":1717497167}}}},"__tags":{"partition_id":"mrigan-test"}}
```

Commit:
gaurds for empty vectors

ArturSztuc committed:
Merge pull request #306 from DUNE-DAQ/asztuc/trigger_tpbuffer_gone

Removed the TP buffers from the trigger

ArturSztuc committed:
Bumping version to 1.8.3

ArturSztuc committed:
Merge pull request #307 from DUNE-DAQ/asztuc/version_bump_v1.8.3

Bumping version to 1.8.3

MRiganSUSX committed:
Merge branch 'mrigan/latencies' into latencies

MRiganSUSX committed:
online latency improvement

MRiganSUSX commented:
@ArturSztuc 

I think all changes are now in. 
The new ones are mostly aimed at offline testing where the offset causes some 'unpleasant behaviour', but should not affect data when running 'live'. 
Also brought in changes from v4.4.2. Should also mention that Grafana plots are now in place as well.

Tested with:
- asset file
- replay with recent tp file
- online with np04 (but this was before the plots were made)
- passes: minimal_system_quick_test, fake_data_producer_test, 3ru_1df_multirun_test

ArturSztuc commented on plugins/ModuleLevelTrigger.cpp:
```diff
@@ -140,6 +148,12 @@ ModuleLevelTrigger::do_configure(const nlohmann::json& confobj)
+  // Latency
+  m_use_latency_monit = params.enable_latency_monit;
+  m_use_latency_offset = params.use_latency_offset;
+  TLOG_DEBUG(TLVL_DEBUG_INFO) << "[MLT] Use latency monitoring?: " << m_use_latency_monit;
+  TLOG_DEBUG(TLVL_DEBUG_INFO) << "[MLT] Use latency offset?: " << m_use_latency_offset;
```
I think we can be probably more liberal with logging in the init/conf, since they are only run once. We could just change all the `_INFO` in them to whichever level outputs everything always.

Not for this PR tho!

ArturSztuc commented on plugins/ModuleLevelTrigger.cpp:
```diff
@@ -330,14 +355,30 @@ ModuleLevelTrigger::send_trigger_decisions()
-                      << tc->time_candidate + m_readout_window_map[tc->type].second;
+
+      if (m_first_tc) {
+        if (m_use_latency_monit && m_use_latency_offset) {
+          m_initial_offset = (duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count()) -
+                             (tc->time_start * m_clock_ticks_to_ms);
+        }
+        m_first_tc = false;
```
This renders replay application somewhat useless for latency measurements.
TC offset will be very different from e.g. TA offset, of from the very first TP offset in the channel filter.
That means we cannot really compare MLT latencies against TAMaker/TCMaker/TPChannelfilter latencies with the replay application.

The same offset should be set for all the applications, probably based off the very first TP's start-time, for replay an emulation.

Perhaps not for this PR though, because it will take too long... but we need to remember this if we want to ever do latencies measurements with replay.

ArturSztuc commented on plugins/ModuleLevelTrigger.cpp:
```diff
@@ -450,18 +494,29 @@ ModuleLevelTrigger::call_tc_decision(const ModuleLevelTrigger::PendingTD& pendin
+          m_system_time =
+            std::chrono::duration_cast<std::chrono::milliseconds>(system_clock::now().time_since_epoch()).count();
+          m_td_send_vs_ro_start.store(
+            fabs(m_system_time - decision.components.back().window_begin * m_clock_ticks_to_ms - m_initial_offset));
+          m_td_send_vs_ro_end.store(
+            fabs(m_system_time - decision.components.back().window_end * m_clock_ticks_to_ms - m_initial_offset));
```
At the moment all the components have exactly the same window, so this is fine, at least for this PR..

I don't know how to resolve it for the future tho. Check if all the components have the same window? What if they do not? Store the window in a separate struct, and have "decision" as part of that struct?

ArturSztuc commented on plugins/TPChannelFilter.cpp:
```diff
@@ -104,21 +118,22 @@ TPChannelFilter::channel_should_be_removed(int channel) const
 
-    std::optional<TPSet> tpset = m_input_queue->try_receive(m_queue_timeout);;
+    std::optional<TPSet> tpset = m_input_queue->try_receive(m_queue_timeout);
+    ;
```
```suggestion
    std::optional<TPSet> tpset = m_input_queue->try_receive(m_queue_timeout);
```
...

ArturSztuc commented on plugins/TPChannelFilter.cpp:
```diff
@@ -136,14 +151,33 @@ TPChannelFilter::do_work()
+        m_system_time =
+          std::chrono::duration_cast<std::chrono::milliseconds>(system_clock::now().time_since_epoch()).count();
+        m_data_vs_system_time.store(fabs(m_system_time - (tpset->start_time * m_clock_ticks_to_ms) - m_initial_offset));
+        TLOG_DEBUG(TLVL_DEBUG_ALL) << "[TPCF] "
+                                   << fabs(
+                                        (m_system_time - (tpset->start_time * m_clock_ticks_to_ms) - m_initial_offset));
```
```suggestion
        TLOG_DEBUG(TLVL_DEBUG_ALL) << "[TPCF] "
                                   << fabs((m_system_time - (tpset->start_time * m_clock_ticks_to_ms) - m_initial_offset));
```
That clang formatting script isn't that great...

ArturSztuc commented on src/trigger/TriggerGenericMaker.hpp:
```diff
@@ -91,12 +95,31 @@ class TriggerGenericMaker : public dunedaq::appfwk::DAQModule
+  }
+  void update_latency_out(uint64_t timestamp)
+  {
+    TLOG_DEBUG(TLVL_DEBUG_ALL) << "TGM OUT: " << timestamp;
```
```suggestion
    TLOG_DEBUG(TLVL_DEBUG_ALL) << "[TGM] OUT: " << timestamp;
```

ArturSztuc commented on src/trigger/TriggerGenericMaker.hpp:
```diff
@@ -91,12 +95,31 @@ class TriggerGenericMaker : public dunedaq::appfwk::DAQModule
+
+  void update_latency_in(uint64_t timestamp)
+  {
+    TLOG_DEBUG(TLVL_DEBUG_ALL) << "TGM IN: " << timestamp;
```
Shouldn't this be:
```suggestion
    TLOG_DEBUG(TLVL_DEBUG_ALL) << "[TGM] IN: " << timestamp;
```

ArturSztuc reviewed (approved):
For the future, let's make clang-formatting across many files a separate PR, to make it easier to review.

There are some general comments, but none that would block this PR (I also made some suggestions on clang formatting mistakes made by the script).

The most major one is that we need to be very careful when using replay to get latencies out, since the time offset applies separately for each DAQModule based on it's first input, rather than having global identical offset.

Other than that, LGTM.

MRiganSUSX committed:
Update plugins/TPChannelFilter.cpp

Co-authored-by: ArturSztuc <a.sztuc@ucl.ac.uk>

MRiganSUSX committed:
Update plugins/TPChannelFilter.cpp

Co-authored-by: ArturSztuc <a.sztuc@ucl.ac.uk>

MRiganSUSX committed:
Update src/trigger/TriggerGenericMaker.hpp

Co-authored-by: ArturSztuc <a.sztuc@ucl.ac.uk>

MRiganSUSX committed:
Update src/trigger/TriggerGenericMaker.hpp

Co-authored-by: ArturSztuc <a.sztuc@ucl.ac.uk>
